### PR TITLE
added vector solutions and space

### PR DESCRIPTION
**Pull request by arthbous**
I added the first layer. This way there is minimal changes to do to each benchmarks (just one line).

Take a look and let me know if you like the direction. Also, the mesh refinement will have to change to take into account the function_space vector.

### CODE REVIEW

**Comment by arthbous** [comment]
@maximilianmetti , running into some issues with Linear_PNP_NS::fasp_solve() in pnp_stokes_eperiment/linear_pnp_ns.cpp 
on how to copy the eigenvector solution to the vector of functions...
My first idea was to copy the vector solution 4 times which seems too much, there must be way using the dofs. Do you know if the dofs are done in order ?

**Comment by maximilianmetti** [comment]
@arthbous Here's where the DoF's are organized by subfunctions.  I haven't used this for vector functions before, but it might give you a head start:
https://github.com/thepnpsolver/modular-pnp/blob/93dfa392a452dae81b14c8c9bb40e511749de3bb/src/pde.cpp#L364

File: benchmarks/pnp_stokes_experiment/linear_pnp_ns.cpp [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+    Function dU = update[3];
+    Function dPressure = update[4];
+
+    std::vector<std::shared_ptr<const Function>> vvv = {dAnion,dCation,dPotential};
+    dolfin::Function update_pnp(_functions_space[0])
+    assign(update_pnp , vvv);
+
+    *(solution[0].vector()) += *(update_pnp.vector());
+    *(solution[1].vector()) += *(dU.vector());
+    *(solution[2].vector()) += *(dPressure.vector());
```
Comment (arthbous) [inline benchmarks/pnp_stokes_experiment/linear_pnp_ns.cpp]: Here is copies the solutions too many times. My questions is: are the DOF in the same order for the function space and the sub function space. Is the following true ?
`dolfin::Function F1(Subspace);
dolfin::Function F2(BigSpace);
F1.vector()[i] = F2.vector[dof[i]]
`

File: src/pde.cpp [hunk 2]
```diff
@@ -151,8 +172,64 @@ void PDE::set_solution (
...
   }
 
-  _bilinear_form->set_coefficient("uu", _solution_function);
-  _linear_form->set_coefficient("uu", _solution_function);
+  _bilinear_form->set_coefficient(_variable, _solution_function);
+  _linear_form->set_coefficient(_variable, _solution_function);
+}
+//--------------------------------------
+void PDE::set_solutions (
+  std::vector<Linear_Function> expression
```
Comment (arthbous) [inline src/pde.cpp:185]: We need a way to have 
`void PDE::set_solutions (std::vector<RandomExpression> expression )`
If we have std::vector<dolfin::Expression> expression as an argument can we still call it with our own Expression class ?

Comment (arthbous) [reply src/pde.cpp:185]: I added a new constructor. I think now it's fine.

**Comment by arthbous** [comment]
Almost done, except the Newton's method does not converges.

@maximilianmetti Also I think [void PDE::set_DirichletBC](https://github.com/thepnpsolver/modular-pnp/pull/76/files#diff-c9e1c964bf47d9086a8bfb5f6112d141R445) and [ void PDE::add_DirichletBC](https://github.com/thepnpsolver/modular-pnp/pull/76/files#diff-c9e1c964bf47d9086a8bfb5f6112d141R467) should be move to linear_PNP.cpp

File: src/pde.cpp [hunk 3]
```diff
@@ -151,8 +177,56 @@ void PDE::set_solution (
...
+
+  std::size_t dimension = new_solutions.size();
+
+  if (_solution_functions.size() == dimension) {
+
+    for (std::size_t i = 0; i < dimension; i++) {
+      _solution_function.reset(new dolfin::Function(_functions_space[i]));
+      *(_solution_functions[i])= new_solutions[i];
+      // _bilinear_form->set_coefficient(_variables[i], _solution_functions[i]);
+      _linear_form->set_coefficient(_variables[i], _solution_functions[i]);
```
Comment (arthbous) [inline src/pde.cpp]: @maximilianmetti In PNP+Stokes some coefficients/variables do not appear in the bilinear form. We need to separate the functions

Comment (maximilianmetti) [reply src/pde.cpp]: @arthbous this is the case with most pdes.  There are two `map`s  for setting the coefficients: https://github.com/thepnpsolver/modular-pnp/blob/develop/src/pde.cpp#L206

`coefficient`s are set for both the bilinear and the linear forms, whereas the `source`s are only set for the linear form.

Comment (arthbous) [reply src/pde.cpp]: That's for Coefficient and Source which is fine. But the PNP+Stokes cases, the variables are separated and the pressure does not appear in the bilinear form. To make it more general we might want to have two variables mapping one for the bilinear form, one for the linear form like Coeff and Source.

Comment (maximilianmetti) [reply src/pde.cpp]: If a variable goes to both forms, it should be in `coefficients`.  If it only goes to the linear form, it goes into `sources`.  Isn't that all you need?

Comment (arthbous) [reply src/pde.cpp]: No. Coefficient and Source are fine. The problem goes back to the separation of variables for PNP+Stokes. With pnp yyou only have a. uu = solution, but with PNP+Stokes some part of the solution does no appear in a.

File: src/pde.cpp [hunk 4]
```diff
@@ -442,11 +537,10 @@ dolfin::Function PDE::_convert_EigenVector_to_Function (
 ) {
   dolfin::Function fn(_function_space);
 
-  if (eigen_vector.size() != _solution_function->vector()->size()) {
-    printf("Cannot convert EigenVector to Function...\n");
-    printf("\tincompatible dimensions!\n");
-  }
-
+  // if (eigen_vector.size() != _solution_function->vector()->size()) {
```
Comment (arthbous) [inline src/pde.cpp:562]: Are you ok removing this line? because in the case of PNP+Stokes '_solution_function" does not exist.

**Review by arthbous — COMMENTED** [review]
It's all working now.

**Comment by arthbous** [comment]
Now we have two options:
1) We merge this and I create a new branch to do a separable PNP+Stokes
2) Or I add it to this branch

Also, for PNP+Stokes seprable which means 4 ufl files, pde.cpp will have to change a lot from one linear form to 4...

**Comment by maximilianmetti** [comment]
@arthbous I want to make sure that this PR will play nicely with the other benchmark problems.

I think that it is much better to go with option 1.  Let's merge this and make a separate branch; otherwise, this PR will be hard to review if we add more.

File: benchmarks/pnp_experiment/main.cpp [hunk 5]
```diff
@@ -139,7 +139,8 @@ int main (int argc, char** argv) {
     pnp_coefficients,
     pnp_sources,
     itsolver,
-    amg
+    amg,
+    "uu"
```
Comment (maximilianmetti) [inline benchmarks/pnp_experiment/main.cpp:143]: It seems that this change will be needed for all benchmarks, correct?

File: src/dirichlet.cpp [hunk 6]
```diff
@@ -57,7 +74,18 @@ void Linear_Function::eval (
   dolfin::Array<double>& values,
   const dolfin::Array<double>& x
 ) const {
-    values[0] = _lower_value * (_mesh_max - x[_coordinate]) * _distance;
-    values[0] += _upper_value * (x[_coordinate] - _mesh_min) * _distance;
+    if (_dimension==1)
```
Comment (maximilianmetti) [inline src/dirichlet.cpp:77]: Why is this conditional needed? If `_dimension == 1`, then the forloop in the `else` statement will only run for `i = 0`.

**Review by maximilianmetti — COMMENTED** [review]
I still need to test this out with other benchmarks and make sure that I can runt the diode benchmarks with this update. 👍

**Comment by arthbous** [comment]
@maximilianmetti Before we merge this one. I should add the benchmark with Lorenz force.

**Comment by maximilianmetti** [comment]
I think that could be a separate PR just to keep the diff small. What do you think?

**Comment by arthbous** [comment]
Need the following changes:
* in error.cpp line 53
 error = error - *_exact_solution; => *error.vector() -= *_exact_solution->vector();
* In Cmakelist.txt need to change ${boost_librariry} to  ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY})
* Now we need to add ${UMFPACK_LIBRARY} with 
set(CMAKE_MODULE_PATH /usr/local/faspsolver/modules)
find_package(UMFPACK REQUIRED)
include_directories(${UMFPACK_INCLUDE_DIRS})

I think modifying EAFE.h could be easy (I can do it). And with All these changes pnp_stokes_experiment works !

P.S. : I hate all the people that update libraries

**Comment by maximilianmetti** [comment]
Are all these changes related to the FEniCS update?  Do you want to get that working and I can see if it's backwards compatible with FEniCS 2016.1?

**Comment by arthbous** [comment]
Don't know. But they should work with either version of FENiCS or boost. It's just that now you have be more clear when linkin libraries.

**Comment by maximilianmetti** [comment]
👍

**Comment by arthbous** [comment]
I pushed the changes. Could you just check your regular pnp still works

**Comment by maximilianmetti** [comment]
It seems to work 👍  
I tested `pnp_refine`, `pnp_diode`, and `pnp_stokes_experiment`.
Feel free to merge whenever you're ready.

**Comment by arthbous** [comment]
Added PNP+Stokes without tensor and also refinement.
All ready to merge now.

**Pull request merged**